src/librustc_resolve/late.rs

```diff
@@ -1536,20 +1536,17 @@ impl<'a, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
         let is_syntactic_ambiguity = !has_sub && bm == BindingMode::ByValue(Mutability::Not);
 
         match res {
-            Res::Def(DefKind::Ctor(_, CtorKind::Const), _)
-            | Res::Def(DefKind::Const, _)
-            | Res::Def(DefKind::ConstParam, _)
-                if is_syntactic_ambiguity =>
-            {
+            Res::Def(
+                DefKind::Ctor(_, CtorKind::Const) | DefKind::Const | DefKind::ConstParam,
```

Comment:
Could you add `Res::SelfCtor` to this list instead?
`Self` is an entirely reasonable const pattern (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8eea5be549cc8d5ae53354d21ab90d3c), and the fact that it's generally interpreted as an `PatKind::Path` and doesn't appear here is a parser's implementation detail.

Reply:
Also, `delay_span_bug` is a hack and I never use it in code I maintain.

Reply:
Done. (It does regress diagnostics a bit.)